### PR TITLE
Feature 2nd milestone

### PR DESCRIPTION
**Pull request by ricardo123321**
started the friendship model, updated controllers and added some friend helpers to our project, we associated the friendships with the users and also implemented the unit tests

### CODE REVIEW

**Review by wathigo — CHANGES_REQUESTED** [review]
### Changes Required
Hello team :wave: 
Good job so far!
Let's make a couple of changes to our friendship feature.

**REQUIRED**
Let's work on our views in order to meet these requirements :point_down: 
* User can invite friends ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/send_invitation.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* User can see friendships requests ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/pending_request.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* User has to accept friendship request to start friendship ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
*  User can reject a friendship request ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/accept_reject_friendship.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* There is a button to invite user to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* Posts timeline includes posts written by logged-in user and their friends ([screenshot]
(https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))
* We have only added tests for friendship model. Let's add unit and integration tests for the friendship feature using Rspec and Capybara([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/))

**After working on the changes, submit another request for review.**

**Happy Hacking :computer: 
Simon Wathigo.**

**Comment by simandebvu** [comment]
### Added 💻 
- Document with ERD Diagram to plan friendship implementation.
- User can invite friends.
- User can see friendship requests.
- User can now accept and reject friendship requests.
- A button to invite user to friendship on users list page and on single user page.
- Posts to be seen by friends only.
- Add feature tests.

File: app/controllers/friendships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+class FriendshipsController < ApplicationController
+  def index
+    @friends = current_user.friendships.all
+  end
+
+  # def new
+  #   @friendship = Friendship.new
```
Comment (uryelah) [inline app/controllers/friendships_controller.rb]: - [x] Please remove commented out code.

File: app/controllers/friendships_controller.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    if @friendship
+      @friendship.destroy
+      redirect_to users_path, notice: 'Succesfully unfriended.'
+    else
+      redirect_to root_path, notice: 'You cant do this.'
+    end
+  end
+
+  # def update
+  #   @user = User.find(params[:format])
```
Comment (uryelah) [inline app/controllers/friendships_controller.rb]: - [x] Make sure the confirm friendship route is here and not in the UsersController, and remove the commented out code.

**Review by uryelah — CHANGES_REQUESTED** [review]
Hi **Ricardo Vera Trallero & Shingi Mandebvu**, I'm Sarah your reviewer for this PR 👩‍🚀 

Good job with your fixes and tests, there are still some issues to fix but I'm sure it will be a breeze for you!🦾

# Required Changes 🏗️ 🐛 💡

- [x] You need to update the  PostsController#timeline_posts so it also gets posts from the user friends, not just from the user.


-----

 **Kindly fix the pointed out issues and comments below before requesting the next code review**

------------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

**Comment by simandebvu** [comment]
## Fixed  🧑‍🔧 💻  

- Posts Controller#timeline_posts to also get posts from users friends and not just the user.
- Removed Code Comments (Cleaner Code).
- Moved the accept friendship action to the friendship controller.

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team     
#### Great Job so far, but there are a few more things required. 
### General
- [x] Everything works fine,but the DRY rule should be respected by using the code like the one below and it should be moved to user model.
```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends + self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```
- You can call it this way `current_user.friends_and_own_posts`
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by simandebvu** [comment]
### Changed 🧑‍🔧  

- DRY principle - PostsController to use a scope instead of two similar push operations to the same array.

**Comment by Anna-Myzukina** [comment]
# [APPROVED]  :heavy_check_mark: 

## Thanks for applying the changes  :+1:

### Go and merge to development.

# :raising_hand:  See you at the next projects.

Happy coding  <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

**Comment by simandebvu** [comment]
Thank you 😄

**Pull request merged**